### PR TITLE
Refacotring

### PR DESCRIPTION
**Pull request by Morganamilo**
Implements #4 partly.

I'm thinking if the client should be an ARC internally as the user may want to have multiple clients dotted around the place.

### CODE REVIEW

**Comment by Morganamilo** [comment]
On top of that, maybe the builders could then take an Arc of the instead of a literal reference. This would eliminate the lifetime but builders don't really need to be long lived anyway.

**Comment by Morganamilo** [comment]
Also realised the tests are failing. It ran into that issue but thought I fixed it. Hmmm.

File: src/options.rs [hunk 1]
```diff
@@ -405,19 +405,7 @@ impl ClientOptions {
...
-        // initialize the final client with auth to pass back to the caller.
-        let rest_client_no_auth =
-            rest::Client::new(http_client.clone(), self.clone(), rest_url.clone());
-
-        let auth = auth::Auth::new(rest_client_no_auth, self.clone());
-
-        let rest_client_with_auth =
-            rest::Client::new_with_auth(http_client, self.clone(), rest_url, auth.clone());
-
-        Ok(rest::Rest::new(auth, rest_client_with_auth, self.clone()))
```
Comment (lmars) [inline src/options.rs:420]: Very happy to see this go!

File: src/rest.rs [hunk 2]
```diff
@@ -6,28 +6,41 @@ use regex::Regex;
...
 
 /// A client for the [Ably REST API].
 ///
 /// [Ably REST API]: https://ably.com/documentation/rest-api
 #[derive(Debug)]
 pub struct Rest {
-    pub auth: auth::Auth,
-    pub channels: Channels,
-    pub client: Client,
+    pub channels: (),
```
Comment (lmars) [inline src/rest.rs:19]: Is this needed?

**Review by lmars — COMMENTED** [review]
I see the tests are failing, but this looks great so far!

File: src/http.rs [hunk 3]
```diff
@@ -254,6 +261,8 @@ impl<T: PaginatedItem, U: PaginatedItemHandler<T>> PaginatedRequestBuilder<T, U>
         // the first request returns an error which would be Some(Err(err)), so
         // we unwrap the Option with a generic error which we don't expect to
         // be encountered by the caller.
+        //
+
```
[the remark this answers is not in the document]
Comment (Morganamilo) [reply src/http.rs:265]: Not sure. May have just been a mistake honestly.

File: src/rest.rs (hunk 2 above)
Comment (Morganamilo) [reply src/rest.rs:19]: So I wasn't sure if there was going to be a need for some sort of channels object (HashMap<String, Channel> or something). The spec points out the channel.get function should return existing channels https://ably.com/docs/client-lib-development-guide/features#RSN2. So I'm assuming there is need for a channel field but I just don't know the type yet.

**Comment by Morganamilo** [comment]
If you could approve this for me. I've got work on top of this as to avoid blocks I just continue with work after each PR. I'll fix the tests and stuff after the next PR.

**Comment by Morganamilo** [comment]
If you could approve this for me. I've got work on top of this as to avoid blocks I just continue with work after each PR. I'll fix the tests and stuff after the next PR.

File: src/rest.rs (hunk 2 above)
Comment (lmars) [reply src/rest.rs:19]: Ok thanks. For the record, I dislike the `channels` object from the spec 🙂

**Review by lmars — APPROVED** [review]

**Comment by Morganamilo** [comment]
Thanks!

**Pull request merged**